### PR TITLE
feature/adding event hubs output plugin

### PR DESCRIPTION
**Pull request by polatengin**
Allow using Azure Event Hubs as an output target, so you can use `--output eventhubs` to fork k6 outputs to Azure Event Hubs.

### CODE REVIEW

File: stats/eventhubs/config.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
+*
+* k6 - a next-generation load testing tool
+* Copyright (C) 2017 Load Impact
```
Comment (imiric) [inline stats/eventhubs/config.go:4]:
```suggestion
* Copyright (C) 2020 Load Impact
```

File: stats/eventhubs/config.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	BufferEnabled    null.Bool          `json:"buffer_enabled,omitempty" envconfig:"K6_EVENTHUBS_BUFFER_ENABLED"`
+	PushInterval     types.NullDuration `json:"push_interval,omitempty" envconfig:"K6_EVENTHUBS_PUSH_INTERVAL"`
+}
+
+// NewConfig creates a new Config instance with default values for some fields.
+func NewConfig() Config {
+	return Config{
+		ConnectionString: null.StringFrom(""),
+		BufferEnabled:    null.BoolFrom(true),
+		PushInterval:     types.NewNullDuration(1*time.Second, false),
```
Comment (imiric) [inline stats/eventhubs/config.go:42]: We should keep the `Valid` value consistent for these null types, so maybe:

```suggestion
		ConnectionString: null.NewString("", false),
		BufferEnabled:    null.NewBool(true, false),
		PushInterval:     types.NewNullDuration(1*time.Second, false),
```

I would prefer to see a usable default value for `ConnectionString`, but apparently Event Hubs can't be started locally, so I guess this is fine.

File: stats/eventhubs/config.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	}
+}
+
+// Apply saves config non-zero config values from the passed config in the receiver.
+func (c Config) Apply(cfg Config) Config {
+
```
Comment (imiric) [inline stats/eventhubs/config.go:48]:
```suggestion
```

File: stats/eventhubs/config_test.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2016 Load Impact
```
Comment (imiric) [inline stats/eventhubs/config_test.go:4]:
```suggestion
 * Copyright (C) 2020 Load Impact
```

File: stats/eventhubs/config_test.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+	"testing"
+
+	"github.com/loadimpact/k6/lib/types"
+	"github.com/stretchr/testify/assert"
+	"gopkg.in/guregu/null.v3"
+)
+
+func TestNewConfig(t *testing.T) {
+	config := NewConfig()
+	assert.Equal(t, sql.NullString(sql.NullString{String: "", Valid: true}), config.ConnectionString.NullString)
```
Comment (imiric) [inline stats/eventhubs/config_test.go:34]: No need to import `sql` here, no?

```suggestion
	assert.Equal(t, null.NewString("", false), config.ConnectionString)
```

**Review by imiric — CHANGES_REQUESTED** [review]
Thanks for proposing this @polatengin! I left a few minor comments, and the testing situation should be improved before we consider merging this.

Also make sure to run `go mod vendor` / `go mod tidy` and commit the changes, otherwise you won't get a passing build in CI.

The k6 team is not familiar with Azure Event Hubs, but from what I could gather [it exposes a Kafka-compliant endpoint](https://docs.microsoft.com/en-us/azure/event-hubs/event-hubs-quickstart-kafka-enabled-event-hubs). Is there a reason why the current [Kafka output](https://k6.io/docs/results-visualization/apache-kafka) or the new [Kafka extension](https://github.com/mostafa/xk6-kafka) couldn't be used instead of adding a custom collector for it?

It also appears to be based on [AMQP](https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-protocol-guide), which is a much more generic and widely used protocol. From that perspective it would be preferable if we added AMQP support instead (say, as a new k6 extension), which would close #1073, and supported Event Hubs that way.

I'm interested to hear your thoughts on this, and also from @na-- @MStoykov.

**Comment by na--** [comment]
I don't have the time to look at this PR closely right now, sorry... I am also not very familiar with Azure Event Hubs or Application Insights (https://github.com/loadimpact/k6/pull/1757). But, in general, if we can use a generic open protocol to output metrics to some service, there needs to be a good reason not to use it and instead to use a custom SDK for it (https://github.com/Azure/azure-event-hubs-go in this case, from what I can see). Moreover, that dependency itself [seems to depend](https://github.com/Azure/azure-event-hubs-go#using-event-hubs) on a [dead library](https://github.com/vcabbage/amqp#packagamqp)... :confused: 

In any case, I also wanted to say that we plan to majorly refactor how the k6 outputs are handled, currently targeting k6 v0.31.0 (Feb 2021). We aim to fix a bunch of issues (https://github.com/loadimpact/k6/issues/1606, https://github.com/loadimpact/k6/issues/1075, https://github.com/loadimpact/k6/issues/1423 among others) and to also enable the addition of output extensions with [xk6](https://github.com/k6io/xk6). So it's very unlikely that we'll merge this PR or https://github.com/loadimpact/k6/pull/1757 before k6 v0.31.0. And even then, I can't promise that these outputs would be merged straight into the k6 core, they might be better off as extensions for a while, to mature. We'd need to gauge the user interest against the maintenance burden of such changes.

**Comment by simskij** [comment]
I can assist here if it's any help, @na--. I'm quite familiar with Azure 👍

**Comment by polatengin** [comment]
> I can assist here if it's any help, @na--. I'm quite familiar with Azure 👍

Hi @simskij thanks for your support offer. It's greatly appreciated 👍

**Comment by polatengin** [comment]
> In any case, I also wanted to say that we plan to majorly refactor how the k6 outputs are handled, currently targeting k6 v0.31.0 (Feb 2021). We aim to fix a bunch of issues (#1606, #1075, #1423 among others) and to also enable the addition of output extensions with [xk6](https://github.com/k6io/xk6). So it's very unlikely that we'll merge this PR or #1757 before k6 v0.31.0

Hey @na-- , is there an `Early Access` program for the new output system with xk6? I'd like to be in such program if you have one, also please count me in if you're going to do some design sessions and looking for community contributions.

**Comment by imiric** [comment]
@polatengin There's no "Early Access" program for xk6, but you're welcome to try creating an extension and publishing it on GitHub. Proper documentation is still pending, but you can take a look at [this article](https://k6.io/blog/extending-k6-with-xk6), and use [existing extensions](https://github.com/topics/xk6) as examples. It's really quite straightforward :)

If you need any help or have any suggestions feel free to post on [the forum](https://community.k6.io/) or [Slack](https://k6.io/slack).

**Comment by MStoykov** [comment]
Hi @polatengin , thanks for this PR :bow:. 

k6 has had support for output extensions for some time now. As of today it also has a [template repo for output extensions](https://github.com/grafana/xk6-output-template). This plus all the other extension outputs should be enough for people to extract this code in an extension and use it without it being in k6 core.

Given that I am closing this PR.

**Pull request closed without merging**